### PR TITLE
TST: placement of network error catching in s3 tests

### PR DESCRIPTION
**Pull request by jreback**

### CODE REVIEW

**Comment by jreback** [comment]
This I believe exposes some issues that exist with the s3 tests and mocking that have our CI occasionaly fail. I think there is some interaction going on. When run separately these all work.

 ``pytest  pandas/tests/io/parser/test_network.py pandas/tests/io/json/test_pandas.py pandas/tests/io/test_parquet.py -r xXs -k s3 -v --tb=short``

xref #19585 

was trying to remove the resource closing issues (IOW on s3 access we don't close s3).

any ideas here: @jorisvandenbossche @TomAugspurger @maximveksler @mdurant

**Review by TomAugspurger — COMMENTED** [review]
Does running `pytest pandas/tests/io/parser/test_network.py pandas/tests/io/json/test_pandas.py pandas/tests/io/test_parquet.py -r xXs -k s3 -v --tb=short` fail for you ever?

I've tried a handful of times and haven't gotten a failure yet, and I've never seen it locally. Are the failures on a run using `pytest-xdist`? Some race condition between creating the fixtures and using them?

**Comment by jreback** [comment]
this fails for me locally - i was expecting it to fail in the CI actually 

let me take a closer look

File: pandas/io/common.py [hunk 1]
```diff
@@ -183,7 +183,10 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
 
     Returns
     -------
-    a filepath_ or buffer or S3File instance, the encoding, the compression
+    tuple of (a filepath_ or buffer or S3File instance,
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/io/common.py]: these are not multiple return values from different possibilities for the first

**Comment by jreback** [comment]
So these pass now all pass, but have these warnings. any ideas?
```
Users/jreback/miniconda3/envs/pandas/lib/python3.6/site-packages/fastparquet/dataframe.py:7: FutureWarning: 'pandas.core' is private. Use 'pandas.Categorical'
  from pandas.core.categorical import Categorical, CategoricalDtype
collected 140 items                                                                                                                                                                                                                         

pandas/tests/io/json/test_pandas.py .                                                                                                                                                                                                 [  7%]
pandas/tests/io/test_parquet.py ..                                                                                                                                                                                                    [ 21%]
pandas/tests/io/parser/test_network.py ...........                                                                                                                                                                                    [100%]

============================================================================================================= warnings summary ==============================================================================================================
pandas/tests/io/parser/test_network.py::TestS3::()::test_parse_public_s3_bucket
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52618), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52626), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52629), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52630), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)

pandas/tests/io/parser/test_network.py::TestS3::()::test_parse_public_s3n_bucket
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52632), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)

pandas/tests/io/parser/test_network.py::TestS3::()::test_parse_public_s3a_bucket
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52633), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)

pandas/tests/io/parser/test_network.py::TestS3::()::test_parse_public_s3_bucket_nrows
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52635), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52640), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)
  /Users/jreback/pandas/pandas/io/s3.py:38: ResourceWarning: unclosed <ssl.SSLSocket fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('172.20.10.2', 52645), raddr=('52.216.96.141', 443)>
    filepath_or_buffer = fs.open(_strip_schema(filepath_or_buffer), mode)

pandas/tests/io/parser/test_network.py::TestS3::()::test_parse_public_s3_bucket_chunked
```

File: pandas/tests/io/conftest.py [hunk 2]
```diff
@@ -2,30 +2,34 @@
...
+@pytest.fixture
+def salaries_table(parser_data):
     """DataFrame with the salaries dataset"""
-    path = os.path.join(HERE, 'parser', 'data', 'salaries.csv')
+    path = os.path.join(parser_data, 'salaries.csv')
     return read_table(path)
 
 
-@pytest.fixture(scope='module')
+@pytest.yield_fixture
```
Comment (TomAugspurger) [inline pandas/tests/io/conftest.py]: I think `yield_fixture`s are deprecated. You can just `yield` in a regular fixture.

**Review by TomAugspurger — COMMENTED** [review]
https://github.com/pandas-dev/pandas/blob/a2771089d87afe8104765f12a07a6cd125e532ac/pandas/core/ops.py#L273-L284 looks somewhat similar. I wonder if that's it?

File: pandas/tests/io/conftest.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/io/conftest.py]: hahah didn't know that!

**Comment by rth** [comment]
```diff
@@ -194,7 +197,8 @@ def get_filepath_or_buffer(filepath_or_buffer, encoding=None,
-        return reader, encoding, compression
+        return reader, encoding, compression, True
```
This resulted in a backward incompatible change for `pandas.io.common.get_filepath_or_buffer` between 0.22 and 0.23. I couldn't find anything about it in the release notes either. Unless it's a private function, but nothing in the import path suggests it...

**Pull request merged**